Fix Stock Notice behavior in the Single Product block

octaedro opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Partially addresses https://github.com/woocommerce/woocommerce/issues/53011.

![Screenshot 2024-11-29 at 5 12 11 PM](https://github.com/user-attachments/assets/e3d3fe63-428b-471a-b2a9-4a5680e867e6)

This PR adds code to determine the availability of the stock indicator based on the following criteria:

- **Product Type:**
  - The stock indicator is only displayed for allowed products: simple products and single variations with `stock management` enabled.
  - The indicator will not be shown when `stock management` is disabled and the product is in stock.
  
These changes ensure consistent visibility of the stock indicator in the `Single Product` block in the editor and on the front end.
This PR also adds the component `Stock Indicator` to the inner blocks within the `Add to Cart with Options (experimental)` block.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Having `WooCommerce Beta Tester` installed and activater
1. Go to Tools > WCA Test Helper > Features
2. Enable the features: `blockified-add-to-cart` and `experimental-blocks`
3. Create some different products (Simple, Variable, Grouped, External).
4. Go to Pages > Add New Page
5. Add a `Single Product` block and select a simple product.
6. Add a `Stock Indicator` block inside the `Single Product` block.
7. The stock indicator won't be visible for single products when:
- The `manage stock` is disabled and is in stock
- The `manage stock` is enabled, quantity is 0 and `Allow backorders?` is Allow
8. Now select a variable product.
9. Pick one of the variations from the variable product you chose before.
- The stock indicator won't be visible when:
  - The parent has `manage stock`, quantity is 0 and `Allow backorders?` is `Allow`. And the picked variation doesn't manage stock.
  - The parent don't has `manage stock` and in the variation is `in stock` is selected.
10. The stock indicator won't be visible for `grouped` and `external` products.
11. Verify that no regressions have been introduced.

## UPDATE

For reference, the product page and the page you just created with the `Single Product` block should display the same stock indicator behavior.


<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

octaedro committed:
Add stock notice to single product block

octaedro committed:
Fix isStockVisible

octaedro committed:
Remove empty line

octaedro commented on plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/stock-indicator/constants.tsx:
```diff
@@ -15,3 +15,5 @@ export const BLOCK_DESCRIPTION: string = __(
 	'woocommerce'
 );
+
+export const ALLOWED_PRODUCT_TYPES = [ 'simple', 'variation' ];
```
I just wanted to highlight that, since we are mimicking the frontend behavior, we won't show the `stock indicator` when the product type is `variable`. We show the `stock indicator` for variations with `stock management` enabled.

octaedro committed:
Add changelog

retrofox reviewed (commented):
Why we are showing a link to the product editor here?
Is it something that we'll add for all children blocks?

<img width="767" alt="Screenshot 2024-12-02 at 2 37 17 PM" src="https://github.com/user-attachments/assets/d8ab396f-19ea-468d-a009-f02d08904142">

octaedro committed:
update block and edit

# Conflicts:
#	plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/edit.tsx

octaedro commented:
>Why we are showing a link to the product editor here?
Is it something that we'll add for all children blocks?

I'm not entirely sure why that message is here, but _I think_ the idea might be to allow modification of the stock options for the product. However, I noticed that this wasn't introduced in this PR. Do you think the message is wrong?

octaedro commented:
Thank you @retrofox for your review!
I just rebased this PR. Could you take another look at this PR?

retrofox commented:
> > Why we are showing a link to the product editor here?
> > Is it something that we'll add for all children blocks?
> 
> I'm not entirely sure why that message is here, but _I think_ the idea might be to allow modification of the stock options for the product. However, I noticed that this wasn't introduced in this PR. Do you think the message is wrong?

Well, what the PR introduces is the block that shows the message.
I would like to know why and what the plan is for the rest of the blocks.

AFAIK the block redirects to the block straight to the block editor.
I don't think there's any need to show this here.

retrofox commented on plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/stock-indicator/block.tsx:
```diff
@@ -10,13 +10,15 @@ import {
 import './style.scss';
 import type { BlockAttributes } from './types';
 
+const ALLOWED_PRODUCT_TYPES = [ 'simple', 'variation' ];
```
I think it should be extensible

octaedro replied:
Good call! I made this extensible in the commit [fd74b54](https://github.com/woocommerce/woocommerce/pull/53319/commits/fd74b542854ec6b4135992f76437fdd1312d7df1)

Aljullu replied:
Ups, sorry, I left a comment about this before seeing this thread. Copying it here instead:

> I wonder if we should mark this filter as experimental, at least for now. Or maybe even remove it until we find a use case where it's needed. (I'm sure there are, but I'm a bit wary of adding the filter so early in the development cycle when we hadn't started looking into extensibility yet). What do you think?

octaedro replied:
Good idea! Let's remove the filter for now and add it back later.

octaedro replied:
Done in the commit [180da0a](https://github.com/woocommerce/woocommerce/pull/53319/commits/180da0a13e59948c5edb004f452f5cac846e35da)

retrofox commented on plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/stock-indicator/block.tsx:
```diff
@@ -64,7 +87,7 @@ export const Block = ( props: Props ): JSX.Element | null => {
 	const { product } = useProductDataContext();
 
-	if ( ! product.id ) {
+	if ( ! product.id || ! isStockVisible( product ) ) {
```
How could it impact the block in a different context? Do you?
Should we check with other stakeholders, or are we safe?

retrofox reviewed (approved):
Approving the PR with a few comments.
Thanks, Fernando!

octaedro committed:
Make allowedProductTypes extensible

octaedro committed:
Convert the whitelist

octaedro committed:
Fix product list

Aljullu commented on plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/stock-indicator/block.tsx:
```diff
@@ -56,6 +58,32 @@ const getTextBasedOnStock = ( {
+ * Determines whether the stock indicator should be visible in the editor.
+ * Stock indicator is not visible when:
+ * - Product type is not allowed (variable, grouped, external)
+ * - Product is sold individually
```
Out of curiosity, why are we hiding the Stock indicator if the product is sold individually? A product can be sold individually but still be "out of stock", "on backorder", etc.

octaedro replied:
Yeah, from what I saw in [the issue](https://github.com/woocommerce/woocommerce/issues/53011)'s description, I thought that we wouldn't show the stock indicator when the product is `sold individually`, but then I realized that I was wrong. I fixed it in the commit: [010b5c1](https://github.com/woocommerce/woocommerce/pull/53319/commits/010b5c1657ff4c152588541cff69dff906be29fd) and fixed the method description in the commit [457fcb7](https://github.com/woocommerce/woocommerce/pull/53319/commits/457fcb7fcd0ad531ff644d18d50c83e763183c41)

Aljullu commented on plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/stock-indicator/block.tsx:
```diff
@@ -56,6 +58,32 @@ const getTextBasedOnStock = ( {
+/**
+ * Determines whether the stock indicator should be visible in the editor.
+ * Stock indicator is not visible when:
+ * - Product type is not allowed (variable, grouped, external)
```
For variables products, shouldn't the block be visible but update according to the selected variations? So if the shopper select an "out of stock" variation, the stock indicator would reflect that.

octaedro replied:
Fixed in the commit [1b5acca](https://github.com/woocommerce/woocommerce/pull/53319/commits/1b5acca46162b74cfddd8e0ea6f3b8df181a0639)

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductStockIndicator.php:
```diff
@@ -38,6 +38,29 @@ protected function get_block_type_uses_context() {
+	 * @return array
+	 */
+	protected function product_types_without_stock_indicator() {
+		return apply_filters( 'woocommerce_blocks_product_types_without_stock_indicator', array( 'external', 'grouped', 'variable' ) );
```
I wonder if we should mark this filter as experimental, at least for now. Or maybe even remove it until we find a use case where it's needed. (I'm sure there are, but I'm a bit wary of adding the filter so early in the development cycle when we hadn't started looking into extensibility yet). What do you think?

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductStockIndicator.php:
```diff
@@ -38,6 +38,29 @@ protected function get_block_type_uses_context() {
 		return [ 'query', 'queryId', 'postId' ];
 	}
 
+		/**
```
Nit:
```suggestion
	/**
```

Aljullu reviewed (commented):
Thanks for working on this, @octaedro! I left some questions inline.

octaedro committed:
Fix stock indicator visibility

octaedro committed:
Add js tests

octaedro committed:
Fix method description

octaedro committed:
Update plugins/woocommerce/src/Blocks/BlockTypes/ProductStockIndicator.php

Co-authored-by: Albert Juhé Lluveras <contact@albertjuhe.com>

octaedro committed:
remove filter

Aljullu commented on plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/stock-indicator/block.tsx:
```diff
@@ -56,6 +58,42 @@ const getTextBasedOnStock = ( {
+		! product.is_on_backorder;
+
+	const isStockHiddenForVariation =
+		! product.backorder_notification_enabled &&
```
According to the definition in the endpoint, `is_on_backorder` will already be false if notifications are disabled, so I think `backorder_notification_enabled` isn't needed?

octaedro replied:
We will need to know if we notify the customer or not. 

![Screenshot 2024-12-05 at 4 04 42 PM](https://github.com/user-attachments/assets/2104c160-7a60-4dc6-8f89-1036b5aedeee)

`is_on_backorder` will be true for the options: `Allow` and `Allow, but notify customer`, however we won't show the message when the option `Allow` is selected. We need the extra attribute in order to know whether we should show the stock indicator or not. Same check is done [here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/abstracts/abstract-wc-product.php#L2177).

Aljullu replied:
Oh, good catch! `is_on_backorder` returns true even if backorder notifications are disabled. :thinking: 

To me, this looks like a bug in the Store API endpoint. I think `is_on_backorder` should actually check that (1) the product is on backorder and (2) notifications are enabled.

Otherwise, it feels weird that we are exposing these internal settings to the Store API, which is public and available to shoppers. For example, as a merchant, I might sell a product pretending I have it in stock even if I don't have it, so IMO `is_on_backorder` should only be true if the product is on backorder **and** notifications are enabled. What do you think?

I think it's worth considering because it would allow us to avoid introducing the `backorder_notification_enabled` property.

Aljullu commented on plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/stock-indicator/block.tsx:
```diff
@@ -56,6 +58,42 @@ const getTextBasedOnStock = ( {
+
+	const isInStockHidden =
+		product.is_in_stock &&
+		product.manage_stock === false &&
```
Sorry, I might be getting confused, but why do we need `manage_stock`? I would expect that `is_in_stock`, `is_on_backorder` and `low_stock_remaining` variables would be enough to know whether we need to display the notice or not, no? :thinking:

octaedro replied:
We need `manage_stock` in order to know if we should show the indicator when is in stock, because if the merchant checks the `Stock management` checkbox, the stock indicator will be visible even when `is_in_stock` is `true`. If that is not selected and `is_in_stock` is `true` we won't show the indicator.

![Screenshot 2024-12-05 at 4 16 56 PM](https://github.com/user-attachments/assets/61b71b21-8898-4549-a90a-7b1af75aaa5b)

Aljullu replied:
If I'm understanding it right, this is something that we will change in https://github.com/woocommerce/woocommerce/pull/53462, no? Asking because I think that if we can avoid adding extra properties to the endpoint, I think it's better.

Aljullu reviewed (commented):
Thanks for the updates, @octaedro! I did a round of testing and a few comments below:

When the block isn't rendered, it looks a bit weird, I wonder if we could add a message in some way, or still display "In stock" but add a notice in the sidebar mentioning that the _Stock Notice_ block might not be rendered in certain cases. What do you think?

![imatge](https://github.com/user-attachments/assets/65351d2a-2cb6-4032-90bc-15948a358040)

When a simple product is in stock, I can see the block in the editor but not in the frontend. Is that expected?

Editor | Frontend
--- | ---
![imatge](https://github.com/user-attachments/assets/6e274270-6214-4c9f-9e65-c4e1eb72cc0f) | ![imatge](https://github.com/user-attachments/assets/5a5955f5-949b-4775-9a10-f7f129e39603)

The settings of this product are:

![imatge](https://github.com/user-attachments/assets/38820272-fba5-4fd2-a4fc-d037e7b19457)

octaedro commented:
> When the block isn't rendered, it looks a bit weird, I wonder if we could add a message in some way, or still display "In stock" but add a notice in the sidebar mentioning that the Stock Notice block might not be rendered in certain cases. What do you think?

I like the idea! What do you think about adding that in a follow-up PR? I have the feeling that this PR is already a little big.

> When a simple product is in stock, I can see the block in the editor but not in the frontend. Is that expected?

I couldn't reproduce this. Could you provide more data about your product?
This is what I see with a simple product:

https://github.com/user-attachments/assets/f924305a-1079-46cf-8381-9769f2bd3489

Aljullu reviewed (commented):
Thanks for answering my questions, @octaedro! I left a couple of extra comments/questions. Sorry for being so stubborn, but I would like to make sure we _really need_ to introduce the `backorder_notification_enabled` and `manage_stock` properties to the endpoint because once they are added, we won't be able to remove them because of backwards compatibility.

> I like the idea! What do you think about adding that in a follow-up PR? I have the feeling that this PR is already a little big.

Yes, agree. Let's not increase the scope of this PR. :+1:

octaedro committed:
Add stock_indicator_text

octaedro committed:
Fix productStockIndicator.php

octaedro committed:
Add save null to index

octaedro committed:
fix block.tsx

octaedro committed:
Fix tests

octaedro committed:
Fix lint

octaedro commented on plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/stock-indicator/index.ts:
```diff
@@ -26,6 +26,7 @@ const blockConfig: BlockConfiguration = {
 	attributes,
 	supports,
 	edit,
+	save: () => null,
```
With this line, we will display the `Stock Indicator` block outside the non-blockified `Add to Cart with Options` and fix [this error](https://github.com/woocommerce/woocommerce/issues/53461).

octaedro committed:
use and return get_availability

Aljullu commented on plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/stock-indicator/block.tsx:
```diff
@@ -17,73 +16,30 @@ import type { HTMLAttributes } from 'react';
-					inStock,
-				'wc-block-components-product-stock-indicator--out-of-stock':
-					! inStock,
+				[ `wc-block-components-product-stock-indicator--${ availabilityClass }` ]:
+					availabilityClass,
```
~While by default `availabilityClass` will always have a value, I wonder if we should check it's not empty before using it. Ie, an extension might hook into `woocommerce_get_availability_class` and set it to an empty string, in that case I think it's better not to add the class to the `<div>`, no?~

Never mind, that's already happening. Sorry for the confusion!

Aljullu commented on plugins/woocommerce/src/StoreApi/Schemas/V1/ProductSchema.php:
```diff
@@ -477,6 +497,7 @@ public function get_item_response( $product ) {
 			'is_in_stock'         => $product->is_in_stock(),
 			'is_on_backorder'     => 'onbackorder' === $product->get_stock_status(),
 			'low_stock_remaining' => $this->get_low_stock_remaining( $product ),
+			'stock_availability'  => $product->get_availability(),
```
What do you think about being explicit about the properties we return in `stock_availability`? I worry that if at some point `get_availability()` is updated to introduce a new property, we would expose it in the endpoint unexpectedly.

```PHP
$availability = $product->get_availability(),
...
			'stock_availability'         => (object) [
				'text'   => $availability['text'] ?? '',
				'class'  => $availability['class'] ?? '',
			],
```

This way we also get a protection in case an extension removed `text` or `class` via `woocommerce_get_availability`.

octaedro replied:
Done in [9e4fca4](https://github.com/woocommerce/woocommerce/pull/53319/commits/9e4fca484bd8609d8631ca7d280cdd1263c3a664)

Aljullu reviewed (approved):
Looks great, outstanding job here, @octaedro! :raised_hands: 

I left two comments inline, but pre-approving.

And two questions from the testing steps:

1. I wonder if we should remove references to the _Add to Cart with Options (Experimental)_ block or separate them into a separate section. Given the _Stock Indicator_ block is available in the stable version of WooCommerce, I think this PR can be tested as part of the release where the experimental blocks aren't available, right?
2. I think `Now select a variable product (not a specific variation). The stock indicator won't be visible.` is not true, no? I can see _Out of stock_ when a variable product is out of stock.

octaedro committed:
Modify availability prop

octaedro committed:
Fix lint

octaedro committed:
fix prop declaration

octaedro commented:
Thank you @Aljullu for your review. I just created [an issue](https://github.com/woocommerce/woocommerce/issues/53720) to tackle the follow-up tasks related to the variable products and the stock indicator.
